MLS.Agent.Tests/FactRunWhenJupyterIsInstalledAttribute.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.IO;
+using WorkspaceServer;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace MLS.Agent.Tests
+{
+    [XunitTestCaseDiscoverer("MLS.Agent.Tests.JupyterInstalledTestCaseDiscoverer", "MLS.Agent.Tests")]
+    public class FactRunWhenJupyterIsInstalledAttribute : FactAttribute
+    {
+    }
+
+    public class JupyterInstalledTestCaseDiscoverer : IXunitTestCaseDiscoverer
+    {
+        private readonly IMessageSink messageSink;
+
+        public JupyterInstalledTestCaseDiscoverer(IMessageSink messageSink)
+        {
+            this.messageSink = messageSink;
+        }
+
+        public IEnumerable<IXunitTestCase> Discover(
+            ITestFrameworkDiscoveryOptions discoveryOptions,
+            ITestMethod testMethod,
+            IAttributeInfo factAttribute)
+        {
+            if (testMethod.TestClass.Class.Name.Contains("Integration") && !FileSystemJupyterKernelSpec.CheckIfJupyterKernelSpecExists().Result)
```

Comment:
It's a good general policy to avoid ever using `Result` as it can often result in deadlocks. I would make `CheckIfJupyterKernelSpecExists` non-async and wrap the async call there in a `Task.Run(...).Wait()`, with a timeout.